exist-core/src/main/java/org/exist/backup/restore/AppRestoreUtils.java

```diff
@@ -0,0 +1,180 @@
+                        LOG.warn("Invalid semver in expath repository for " + app.name, e);
+                    }
+                }
+            }
+        }
+        return paths;
+    }
+
+    /**
+     * Inspect all collections which may belong to apps in the backup descriptor. Return a list
+     * of {@link AppDetail} objects containing the symbolic path, name and version of every app
+     * found.
+     *
+     * The method attempts to be fail safe to make sure even bad backups can be restored. Errors
+     * reading package descriptors are thus only logged and should not abort the process.
+     *
+     * @param broker the broker to use for parsing the descriptor and obtaining the app root
+     * @param descriptors a queue of backup descriptors to inspect
+     * @return list of application details
+     */
+    private static List<AppDetail> getAppsFromBackup(DBBroker broker, Deque<BackupDescriptor> descriptors) {
+        final String appRoot = getAppRoot(broker);
+        final List<AppDetail> result = new ArrayList<>(17);
+        final XMLReaderPool parserPool = broker.getBrokerPool().getParserPool();
+        for (final BackupDescriptor descriptor : descriptors) {
+            final BackupDescriptor apps = descriptor.getChildBackupDescriptor(appRoot);
+            if (apps != null) {
+                final List<String> collections = getSubcollectionNames(parserPool, apps);
+
+                for (final String collection : collections) {
+                    final BackupDescriptor app = apps.getChildBackupDescriptor(collection);
+                    final InputSource is = app.getInputSource("expath-pkg.xml");
+                    if (is != null) {
+                        XMLReader reader = null;
+                        try {
+                            reader = parserPool.borrowXMLReader();
+                            reader.setContentHandler(new DefaultHandler() {
+                                @Override
+                                public void startElement(String uri, String localName, String qName, Attributes attributes) throws SAXException {
+                                    if (PKG_NAMESPACE.equals(uri) && "package".equals(localName)) {
+                                        final String version = attributes.getValue("version");
+                                        final String name = attributes.getValue("name");
+                                        if (StringUtils.isEmpty(version) || StringUtils.isEmpty(name)) {
+                                            LOG.warn("Invalid package descriptor for " + app.getSymbolicPath());
+                                            return;
+                                        }
+                                        try {
+                                            final AppDetail detail = new AppDetail(app.getSymbolicPath(), name, Semver.parse(version));
+                                            result.add(detail);
```

Comment:
This will only work for apps which have valid SemVer 2.0 version numbers. @joewiz and I just fixed this elsewhere.
I am not certain, but I would that that could be a concern here too as there are some older packages with non-Semver 2.0 compatible naming?